### ISSUE TITLE
Swarm panics on Engine refreshContainer after TTL node expiration

### ISSUE THREAD

**Issue by abronan** [docker/swarm#1093]
When testing the backend KV failure mechanism and trying a simple `docker run` after a recover, swarm eventually panics on `refreshContainer`. There seems to be a race when the TTL of a node expires after the recovery process and the `refreshContainer` is triggered while the Engine is being removed.

The (reduced) output:

```
INFO[4338] Leader Election: Cluster leadership acquired 
INFO[4340] New leader elected: 192.168.42.201:4000      
INFO[4349] Removed Engine swarm-agent-1  

panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x68 pc=0x558b4d]

goroutine 62 [running]:
github.com/docker/swarm/cluster.(*Engine).refreshContainer(0xc208035e10, 0xc2083b2740, 0x40, 0x0, 0x0, 0x0)
    /vagrant/src/github.com/docker/swarm/cluster/engine.go:231 +0x20d
github.com/docker/swarm/cluster.(*Engine).handler(0xc208035e10, 0xc20837e780, 0x0, 0x0, 0x0, 0x0)
    /vagrant/src/github.com/docker/swarm/cluster/engine.go:547 +0x283
github.com/docker/swarm/cluster.*Engine.(github.com/docker/swarm/cluster.handler)·fm(0xc20837e780, 0x0, 0x0, 0x0, 0x0)
    /vagrant/src/github.com/docker/swarm/cluster/engine.go:111 +0x59
github.com/samalba/dockerclient.func·005()
    /vagrant/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/samalba/dockerclient/dockerclient.go:378 +0x1c8
created by github.com/samalba/dockerclient.(*DockerClient).StartMonitorEvents
    /vagrant/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/samalba/dockerclient/dockerclient.go:380 +0x1b0

goroutine 1 [chan receive, 5 minutes]:
github.com/docker/swarm/api.(*Server).ListenAndServe(0xc2080b6300, 0x0, 0x0)
    /vagrant/src/github.com/docker/swarm/api/server.go:117 +0x491
github.com/docker/swarm/cli.manage(0xc2080900c0)
    /vagrant/src/github.com/docker/swarm/cli/manage.go:273 +0x116f
github.com/codegangsta/cli.Command.Run(0x9f1f30, 0x6, 0x9f1e90, 0x1, 0xa2d310, 0x17, 0x0, 0x0, 0x0, 0x0, ...)
    /vagrant/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/codegangsta/cli/command.go:113 +0x1038
github.com/codegangsta/cli.(*App).Run(0xc2080355f0, 0xc20800a000, 0x8, 0x8, 0x0, 0x0)
    /vagrant/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/codegangsta/cli/app.go:156 +0xcf7
github.com/docker/swarm/cli.Run()
    /vagrant/src/github.com/docker/swarm/cli/cli.go:57 +0x45d
main.main()
    /vagrant/src/github.com/docker/swarm/main.go:13 +0x1b

goroutine 5 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0xcc0320)
    /vagrant/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/golang/glog/glog.go:879 +0x78
created by github.com/golang/glog.init·1
    /vagrant/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/golang/glog/glog.go:410 +0x2a7

goroutine 17 [syscall, 5 minutes, locked to thread]:
runtime.goexit()
    /home/vagrant/.gvm/gos/go1.4/src/runtime/asm_amd64.s:2232 +0x1

goroutine 6 [sleep, 5 minutes]:
github.com/docker/libkv/store/etcd.func·001()
    /vagrant/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/docker/libkv/store/etcd/etcd.go:61 +0x4a
created by github.com/docker/libkv/store/etcd.New
    /vagrant/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/docker/libkv/store/etcd/etcd.go:63 +0x201

goroutine 7 [select]:
github.com/docker/swarm/discovery/kv.(*Discovery).watchOnce(0xc20802e280, 0x0, 0xc2083887e0, 0xc2080db2c0, 0xc2080db320, 0x0)
    /vagrant/src/github.com/docker/swarm/discovery/kv/kv.go:74 +0x575
github.com/docker/swarm/discovery/kv.func·001()
    /vagrant/src/github.com/docker/swarm/discovery/kv/kv.go:118 +0x217
created by github.com/docker/swarm/discovery/kv.(*Discovery).Watch
    /vagrant/src/github.com/docker/swarm/discovery/kv/kv.go:128 +0x1b8

goroutine 8 [select]:
github.com/docker/swarm/cluster/swarm.(*Cluster).monitorDiscovery(0xc2080db260, 0xc2080db2c0, 0xc2080db320)
    /vagrant/src/github.com/docker/swarm/cluster/swarm/cluster.go:238 +0x46d
created by github.com/docker/swarm/cluster/swarm.NewCluster
    /vagrant/src/github.com/docker/swarm/cluster/swarm/cluster.go:55 +0x436

goroutine 24 [chan receive]:
github.com/docker/libkv/store/etcd.(*etcdLock).waitLock(0xc208180740, 0xc20827fca0, 0x1a, 0xc208384ea0, 0xc208384f00)
    /vagrant/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/docker/libkv/store/etcd/etcd.go:484 +0xb6
github.com/docker/libkv/store/etcd.(*etcdLock).Lock(0xc208180740, 0xc2081a2100, 0x0, 0x0)
    /vagrant/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/docker/libkv/store/etcd/etcd.go:446 +0x32c
github.com/docker/swarm/leadership.(*Candidate).campaign(0xc2080db4a0, 0x7fe9d928cfe0, 0xc208180740)
    /vagrant/src/github.com/docker/swarm/leadership/candidate.go:96 +0xa9
created by github.com/docker/swarm/leadership.(*Candidate).RunForElection
    /vagrant/src/github.com/docker/swarm/leadership/candidate.go:56 +0x227

goroutine 1152 [select, 1 minutes]:
net/http.(*persistConn).roundTrip(0xc2083c3600, 0xc20828a9a0, 0x0, 0x0, 0x0)
    /home/vagrant/.gvm/gos/go1.4/src/net/http/transport.go:1082 +0x7ad
net/http.(*Transport).RoundTrip(0xc2080af200, 0xc208328340, 0xc208256420, 0x0, 0x0)
    /home/vagrant/.gvm/gos/go1.4/src/net/http/transport.go:235 +0x558
net/http.send(0xc208328340, 0x7fe9d9286ed8, 0xc2080af200, 0x27, 0x0, 0x0)
    /home/vagrant/.gvm/gos/go1.4/src/net/http/client.go:219 +0x4fc
net/http.(*Client).send(0xc2083ebb00, 0xc208328340, 0x27, 0x0, 0x0)
    /home/vagrant/.gvm/gos/go1.4/src/net/http/client.go:142 +0x15b
net/http.(*Client).doFollowingRedirects(0xc2083ebb00, 0xc208328340, 0xb12fb8, 0x0, 0x0, 0x0)
    /home/vagrant/.gvm/gos/go1.4/src/net/http/client.go:367 +0xb25
net/http.(*Client).Get(0xc2083ebb00, 0xc2082563c0, 0x28, 0x3, 0x0, 0x0)
    /home/vagrant/.gvm/gos/go1.4/src/net/http/client.go:299 +0xba
github.com/samalba/dockerclient.(*DockerClient).MonitorEvents(0xc2083ebb30, 0x0, 0xc20832bf20, 0x0, 0x0, 0x0)
    /vagrant/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/samalba/dockerclient/dockerclient.go:335 +0xbd4
github.com/samalba/dockerclient.func·005()
    /vagrant/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/samalba/dockerclient/dockerclient.go:367 +0x62
created by github.com/samalba/dockerclient.(*DockerClient).StartMonitorEvents
    /vagrant/src/github.com/docker/swarm/Godeps/_workspace/src/github.com/samalba/dockerclient/dockerclient.go:380 +0x1b0
```

**Comment by abronan** [comment]
Re-opening, the issue on `refreshContainer` is still there on https://github.com/docker/swarm/blob/master/cluster/engine.go#L231. We trigger a refresh while `e.client` is nil.

**Comment by abronan** [comment]
Fixed by #1168
